### PR TITLE
Added Shell Sort [Java]

### LINKED ISSUE #391: Shell Sort [Java]
#### This issue is about [Algo/DS Name](link to resource for the Algo/DS).

- [X] Issue Title is in form `Algo/DS Name [Language]`.
- [X] I searched or browsed the repo’s other issues to ensure this is not a duplicate.
- [X] This Algo/DS is worth adding to this repository.
- [X] Assign this issue to me (I want to work on this).
	- [X] I ensure that I am not already assigned to 2 or more issues.
- [x] I will send my PR only after I'm assigned to this issue by any maintainer.
- [X] I understand that I will be unassigned from this issue if I fail to send a PR for this within a week.

### PR DESCRIPTION
**Pull request by jsroyal**
Fixes #391

#### By submitting this pull request I confirm I've read and complied with the below declarations.

- [X] I have read the [Contribution guidelines](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md) and I am confident that my PR reflects them.
- [X] I have followed the [coding guidelines](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md#cs) for this project.
- [X] My code follows the [skeleton code structure](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md#sample).
- [X] This pull request has a descriptive title. For example, `Added {Algorithm/DS name} [{Language}]`, not `Update README.md` or `Added new code`.
- [X] This pull request will be closed if I fail to update it even once in a continuous time span of 7 days.

### CODE REVIEW

**Comment by mohitkyadav** [comment]
@jsroyal As time goes on the number of commits will increase.  Please squash all these 46 commits. Prefer creating a branch on your `algos` repository and send PR from that branch.

**Comment by jsroyal** [comment]
>Please squash all these 46 commits
??

**Comment by mohitkyadav** [comment]
@jsroyal One thing you can do is, save your current work to a temporary branch and perform a hard reset to `upstream/master`.
`git branch temp`
`git reset --hard <tag>/<branch_name>`
ie
`git reset --hard origin/master`
after this your local repo should me even and now you can merge your temporary branch.
 I am not sure if it's safe or we should go with it. Someone, please correct me if I am wrong.

File: shell_sort/ShellSort.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
+// Worst case time complexity = O(n^2)
+// Best case complexity = O(nlog(n))
```
Comment (mohitkyadav) [inline shell_sort/ShellSort.java]: Please add details about `n`.

File: shell_sort/ShellSort.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+public class ShellSort {
+
```
Comment (mohitkyadav) [inline shell_sort/ShellSort.java]: Blank line.

**Comment by Monal5031** [comment]
@mohitkyadav shouldn't it be `git reset --hard upstream/master`?

**Comment by mohitkyadav** [comment]
That's why I wrote ` <tag>/<branch_name>`, `tag` can be `origin` or `upstream` dependent on where you want to reset.

**Comment by Monal5031** [comment]
@mohitkyadav True, but if the PR is made from the master branch itself and the master branch itself has 41 extra commits, thats why I pointed it out to be upstream.

**Comment by jsroyal** [comment]
>$ git branch temp
$ git remote
origin
upstream
$ git reset --hard upstream/master
HEAD is now at 271e7ac Added Counting Sort [JavaScript] (#386)
$ git merge temp
Updating 271e7ac..801cca9
Fast-forward
 README.md                 |  2 +-
 shell_sort/ShellSort.java | 39 +++++++++++++++++++++++++++++++++++++++
 2 files changed, 40 insertions(+), 1 deletion(-)
 create mode 100644 shell_sort/ShellSort.java
$ git status
On branch master
Your branch is up-to-date with 'origin/master'.
nothing to commit, working directory clean

:1234: :4321:

File: shell_sort/ShellSort.java (hunk 1 above)
Comment (jsroyal) [inline shell_sort/ShellSort.java]: >Please add details about n.

??
Number of nodes = n

File: shell_sort/ShellSort.java (hunk 2 above)
Comment (jsroyal) [inline shell_sort/ShellSort.java]: Do you mean to remove -> 3 or -> 5

Comment (mohitkyadav) [inline shell_sort/ShellSort.java]: 3 and 5.

**Comment by mohitkyadav** [comment]
@jsroyal sir after 
>git reset --hard upstream/master
HEAD is now at 271e7ac Added Counting Sort [JavaScript] (#386)

Instead of  `git merge temp`
please do `git merge --squash temp`

File: shell_sort/ShellSort.java (hunk 1 above)
Comment (aviaryan) [inline shell_sort/ShellSort.java]: @jsroyal Yes, you can specify that in the comment.

**Comment by jsroyal** [comment]
jsroyal@HP:~/GitHub/Open Source/algos$ git reset --hard upstream/master
HEAD is now at 271e7ac Added Counting Sort [JavaScript] (#386)
jsroyal@HP:~/GitHub/Open Source/algos$ git merge --squash temp
Updating 271e7ac..801cca9
Fast-forward
Squash commit -- not updating HEAD
 README.md                 |  2 +-
 shell_sort/ShellSort.java | 39 +++++++++++++++++++++++++++++++++++++++
 2 files changed, 40 insertions(+), 1 deletion(-)
 create mode 100644 shell_sort/ShellSort.java
jsroyal@HP:~/GitHub/Open Source/algos$ git status
On branch master
Your branch is behind 'origin/master' by 48 commits, and can be fast-forwarded.
  (use "git pull" to update your local branch)
Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

	modified:   README.md
	new file:   shell_sort/ShellSort.java

jsroyal@HP:~/GitHub/Open Source/algos$ git push origin master
To git@github.com:jsroyal/algos.git
 ! [rejected]        master -> master (non-fast-forward)
error: failed to push some refs to 'git@github.com:jsroyal/algos.git'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
jsroyal@HP:~/GitHub/Open Source/algos$ git pull origin master
From github.com:jsroyal/algos
 * branch            master     -> FETCH_HEAD
Updating 271e7ac..801cca9
Fast-forward
 README.md                 |  2 +-
 shell_sort/ShellSort.java | 39 +++++++++++++++++++++++++++++++++++++++
 2 files changed, 40 insertions(+), 1 deletion(-)
 create mode 100644 shell_sort/ShellSort.java
jsroyal@HP:~/GitHub/Open Source/algos$ git status
On branch master
Your branch is up-to-date with 'origin/master'.
nothing to commit, working directory clean
jsroyal@HP:~/GitHub/Open Source/algos$

**Comment by jsroyal** [comment]
Ok I will take a look on docs.
Thanks BTW.

**Comment by Monal5031** [comment]
@mohitkyadav @jsroyal sir, please do this discussion gitter.

**Comment by jsroyal** [comment]
Anybody got, How to this issue how to re-solve?

**Comment by mohitkyadav** [comment]
@jsroyal hard reset your master branch like before but this time instead of merging your "temp" branch, make changes manually. I think this can resolve this issue.

**Pull request closed without merging**